docs/architecture/architecture.md

```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+request. `io.kubernetes.cri-o.ContainerType` can either be set to `sandbox` or `container`.
+Clear Containers will then use this annotation to decide if it needs to respectively
+create a virtual machine or a container inside a virtual machine associated with a
+Kubernetes pod:
+
+```Go
+	containerType, err := ociSpec.ContainerType()
+	if err != nil {
+		return err
+	}
+
+	switch containerType {
+	case vc.PodSandbox:
+		process, err = createPod(ociSpec, runtimeConfig, containerID, bundlePath, console, disableOutput)
+		if err != nil {
+			return err
+		}
+	case vc.PodContainer:
+		process, err = createContainer(ociSpec, containerID, bundlePath, console, disableOutput)
+		if err != nil {
+			return err
+		}
+	}
+
+```
+
+### Generalization
+
+With Clear Containers 2.1, we implemented the initial Kubernetes/`CRI-O` support by choosing
+to handle pods as a special case for containers. By default the runtime would only be aware
+of containers and only when receiving CRI-O specific annotations it would follow specific
+code paths to handle pods and intra pod containers. This approach showed that container
+orchestration support was an after thought with 2.1, and that supporting the single container
+Docker use case was the main focus.
+
+Starting with Clear Containers 3.0 we took a different, almost opposite approach. The Clear
+Containers 3.0 runtime is based on the `virtcontainers` package which API is inspired by the
+Kubernetes `CRI` one. `virtcontainers` execution units are pods and it does not create or manage
+any container outside of an existing pod context, and thus so does the Clear Container 3.0 runtime.
+To create a Docker container it will first create a pod which will represent the container
+hardware virtualized context and then start an actual container inside that virtual machine.
+Both the pod and its container have the same id and name.
+
+### Mixing VM based and namespace based runtimes
+
+One interesting evolution of the `CRI-O` support for the Clear Containers runtime is the ability
+to run VM based pods alongside with namespace ones. With `CRI-O` and Clear Containers, one can
+introduce the concept of workload trust inside a Kubernetes cluster.
+
+A cluster operator can now tag (through Kubernetes annotations) container workloads as `trusted`
+or `untrusted`. In a software development context, an example of a `trusted` workload would be
+a containerized continuous integration engine whereas all developers applications would be
+`untrusted` by default. Developers workloads can be buggy, unstable or even include malicious
+code and thus from a security perspective it makes sense to tag them as `untrusted`. A `CRI-O`
+and Clear Containers based Kubernetes cluster handles this use case transparently as long as
+the deployed containers are properly tagged. All `untrusted` containers will be handled by
+Clear Container and thus run in a hardware virtualized secure sandbox while `runc` will handle
```

Comment:
"whilst `runc`, for example, could handle ..." maybe to allow for other runtimes?